reactor-netty-core/src/main/java/reactor/netty/FutureMono.java

```diff
@@ -126,20 +113,41 @@ public void subscribe(CoreSubscriber<? super Void> s) {
 				return;
 			}
 
-			if (f.isDone()) {
-				if (f.isSuccess()) {
-					Operators.complete(s);
-				}
-				else {
-					Operators.error(s, FutureSubscription.wrapError(f.cause()));
-				}
-				return;
+			doSubscribe(s, f);
+		}
+	}
+
+	static <F extends Future<Void>> void doSubscribe(CoreSubscriber<? super Void> s, F future) {
+		if (future.isDone()) {
+			if (future.isSuccess()) {
+				Operators.complete(s);
 			}
+			else {
+				Operators.error(s, FutureSubscription.wrapError(future.cause()));
+			}
+			return;
+		}
 
-			FutureSubscription<F> fs = new FutureSubscription<>(f, s);
-			s.onSubscribe(fs);
-			// Returned value is deliberately ignored
-			f.addListener(fs);
+		FutureSubscription<F> fs = new FutureSubscription<>(future, s);
+		// propagate subscription before adding listener to avoid any race between finishing future and onSubscribe
+		// is called
+		s.onSubscribe(fs);
+
+		// check if subscription was not cancelled immediately.
+		if (fs.cancelled) {
+			// if so do nothing anymore
+			return;
+		}
+
+		// add listener to the future to propagate on complete when future is done
+		// addListener likely to be thread safe method
+		// Returned value is deliberately ignored
+		future.addListener(fs);
+
+		// check once again if is cancelled to see if we need to removeListener in case addListener racing with
+		// subscription.cancel (which should remove listener)
+		if (fs.cancelled) {
+			future.removeListener(fs);
```

Comment:
nit: Add comment that the returned value is deliberately ignored

Reply:
done